### ISSUE TITLE
Matrix-Free DG with hanging nodes over periodic faces

### ISSUE THREAD

**Issue by kronbichler** [dealii/dealii#8905]
As reported on the mailing list at https://groups.google.com/forum/#!topic/dealii/okNWVIHwcTA , there is a bug in the setup the matrix free face pair in case we have hanging nodes over periodic faces, i.e., a level difference of one over periodic faces. To reproduce, change the initialization from `step-59` into
```
            Point<dim> upper_right;
            upper_right[0] = 2.5;
            for (unsigned int d = 1; d < dim; ++d)
              upper_right[d] = 2.8;
            GridGenerator::hyper_rectangle(triangulation,
                                           Point<dim>(),
                                           upper_right);
            triangulation.begin_active()->face(0)->set_boundary_id(10);
            triangulation.begin_active()->face(1)->set_boundary_id(11);
            triangulation.begin_active()->face(2)->set_boundary_id(0);
            for (unsigned int f = 3; f < GeometryInfo<dim>::faces_per_cell; ++f)
              triangulation.begin_active()->face(f)->set_boundary_id(1);

            std::vector<GridTools::PeriodicFacePair<
              typename Triangulation<dim>::cell_iterator>>
              periodic_faces;
            GridTools::collect_periodic_faces(
              triangulation, 10, 11, 0, periodic_faces);
            triangulation.add_periodicity(periodic_faces);

            triangulation.refine_global(1);
            triangulation.begin_active()->set_refine_flag();
            triangulation.execute_coarsening_and_refinement();
          
        setup_system();
```
(slightly reduced from the post on the mailing list.)

**Comment by mwichro** [comment]
This is not related periodic faces, I just forgotten to remove some code from step-59. The following code also reproduces the problem:
```
            Point<dim> upper_right;
            upper_right[0] = 2.5;
            for (unsigned int d = 1; d < dim; ++d)
              upper_right[d] = 2.8;
            GridGenerator::hyper_rectangle(triangulation,
                                           Point<dim>(),
                                           upper_right);
            triangulation.begin_active()->face(0)->set_boundary_id(10);
            triangulation.begin_active()->face(1)->set_boundary_id(11);
            triangulation.begin_active()->face(2)->set_boundary_id(0);
            for (unsigned int f = 3; f < GeometryInfo<dim>::faces_per_cell; ++f)
              triangulation.begin_active()->face(f)->set_boundary_id(1);

            triangulation.refine_global(1);

            Vector<float> estimated_error_per_cell(triangulation.n_active_cells());
            for (unsigned int i = 0; i < estimated_error_per_cell.size(); ++i)
              estimated_error_per_cell[i] = std::sin(i) * std::sin(i);

            parallel::distributed::GridRefinement::refine_and_coarsen_fixed_number(
                triangulation, estimated_error_per_cell, 0.3, 0.0);
            triangulation.execute_coarsening_and_refinement();
          }
        setup_system();
```

**Comment by kronbichler** [comment]
You are right - this is related to the setup of level matrix-free objects in the presence of hanging nodes. I'm investigating.

**Comment by kronbichler** [comment]
Closer investigating this topic revealed that we currently do not support the computation of the DG level matrices with the matrix-free framework in the presence of hanging nodes (aka "refinement edges" in the multigrid speak). We did not currently have a good assertion for that, so I started looking in the wrong place. I have added the assertion in #8910. 

The question to you @mwichro is whether you actually need to evaluate the DG level matrices, i.e., if you need to do multigrid, for that case. The reason is that we do not have a good matrix-free interface to work on the crossing between two levels, which is necessary for computing the appropriate DG flux matrices. So that would involve some additional work on the infrastructure and the integrators, given that we would need to both implement homogeneous Dirichlet conditions on the refinement edges for the level smoothing and the flux to the next coarser level when we compute the residual. For that, we would likely need additional abstractions in terms of a matrix-free base operator (as we will also likely need to address #8906), which is likely not done in the next two months.

A solution that I imagine could come earlier is a multigrid infrastructure that lets you transfer from a DG space to a continuous finite element space, where the respective transfer terms are available.

**Comment by mwichro** [comment]
I'm working on a multigrid method for DG formulation of Stokes problem. The MG requires DG formulation, so DG level matrices  are essential for me. I've got very promising results on uniformly refined meshes (also DG matrix-free)

For my method I need homogeneous Dirichlet boundary condition on refinement edges, that  can be also applied by AffineConstrains. So in my case I can manage without computing edge matrices.

**Comment by kronbichler** [comment]
> For my method I need homogeneous Dirichlet boundary condition on refinement edges, that can be also applied by AffineConstrains. 

We could also treat the edge faces as if they were boundaries, so applying homogeneous Dirichlet would not be very much work. The thing I'm concerned about in terms of data structures is that an optimal MG algorithm with those faces needs the flux terms of the residual you restrict to the next coarser level, and similarly the contribution of the flux corrections to the solution from the coarser side during prolongation, in the spirit of the paper by Guido Kanschat from 2004.

**Comment by mwichro** [comment]
I'll look into that  (and step-39 )  and see what I need.  Am I guessing correctly that the missing part correspond to lines:
```
MeshWorker::Assembler::MGMatrixSimple<SparseMatrix<double>> assembler;
assembler.initialize(mg_matrix);
assembler.initialize_fluxes(mg_matrix_dg_up, mg_matrix_dg_down);
```
in step-39? And corresponding assembler for matrix-based program is defined here:
https://dealii.org/developer/doxygen/deal.II/classMeshWorker_1_1Assembler_1_1MGMatrixSimple.html#ad854b648b394a116f6e0a6f9cf712a9b
?

**Comment by kronbichler** [comment]
With #10004, we have a warning for the user that avoids the problem. A proper resolution of this issue is
- some cleanup in the setup of the refinement edges to make their boundary indicator appear in the face loops so a user can add appropriate face integrals with a suitable id,
- to provide a *separate* matrix-free object on refinement edges that specifically takes the coupling terms into account to let a user do the desired faces, so we need to write a specific reinit function; this is also related to #2000 and how we want to proceed with that part, and
- suitable tests that check that all of this is properly working.

This is a bigger task and cannot be done for the 9.2 release.

**Comment by peterrum** [comment]
To rephrase the comment above: This is a bigger task and cannot be done for the **9.3** release.

**Comment by peterrum** [comment]
This one would be nice sometime, but currently I am happy with DG + global coarsening. Adjusting the milestone....
